### PR TITLE
Add metadata_entry attribute to all current resources and data sources to support access levels, domain and types in VCD Metadata

### PR DESCRIPTION
**Pull request by adambarreiro**
### Overview
This PR makes all resources and data sources that support metadata to handle all the attributes of a Metadata Entry (which are: **key**, **value**, **type**, **access level** and **domain**). Before, we lacked from the access level and the domain, and also the type (string, number, boolean, date) was restricted to string.

To do this, I've added the **NEW** attribute `metadata_entry` to all current resources and data sources, which is a **TypeSet** with the mentioned attributes as children. `metadata` is then **deprecated**, and **either** `metadata` or a set of `metadata_entry` is supported in a single resource (ConflictsWith).

This PR depends on https://github.com/vmware/go-vcloud-director/pull/510

### Detailed description

The main changes are centralized in **metadata.go** and **metadata_common_test.go**. These two files provide an abstraction for Metadata implementation in the provider, some sort of "SDK", so adding the metadata support to a new resource should be quite simple:

* Use `getMetadataEntrySchema` to specify the schema.
* Use `createOrUpdateMetadataInVcd` on resource creation and update.
* Use `updateMetadataInState` on resource/data source reads.

For tests, one can use just `testMetadataEntryCRUD(t, resourceHclTemplate, resourceAddress, datasourceHclTemplate, datasourceAddress, extraParams)` like for example:

```go
func TestAccVcdCatalogMediaMetadata(t *testing.T) {
	testMetadataEntryCRUD(t,
		testAccCheckVcdCatalogMediaMetadata, "vcd_catalog_media.test-catalog-media",
		testAccCheckVcdCatalogMediaMetadataDatasource, "data.vcd_catalog_media.test-catalog-media-ds",
		StringMap{
			"Catalog":   testConfig.VCD.Catalog.NsxtBackedCatalogName,
			"MediaPath": testConfig.Media.MediaPath,
		})
}

const testAccCheckVcdCatalogMediaMetadata = `
resource "vcd_catalog_media" "test-catalog-media" {
  org        = "{{.Org}}"
  name       = "{{.Name}}"
  catalog    = "{{.Catalog}}"
  media_path = "{{.MediaPath}}"
  {{.Metadata}}
}
`

const testAccCheckVcdCatalogMediaMetadataDatasource = `
data "vcd_catalog_media" "test-catalog-media-ds" {
  org     = vcd_catalog_media.test-catalog-media.org
  catalog = vcd_catalog_media.test-catalog-media.catalog
  name    = vcd_catalog_media.test-catalog-media.name
}
`
```
To create a test that checks Create, Read, Update and Delete operations for metadata, with all possible types, access levels and domain options.

Nothing more should be needed.

The other changes in the remaining files are just usages of the above.

### Worth mentioning coding decisions

As you can see in **metadata.go** and **metadata_common_test.go**, there are plenty of workarounds. As we require to support old `metadata` attribute with the new `metadata_entry`, there's no other choice but having both `Computed`, because both update a single thing in VCD (metadata entries).

Having the attributes as Computed produce more quirks, described more precisely in the code comments. The most important is, **one can't delete metadata by removing the attribute**. As they're computed, Terraform thinks there is no change and doesn't trigger Update. To force a deletion, one must set:

```hcl
metadata_entry {}
```
This will trigger an Update in the resource and delete metadata. This has a further consequence: We can't have required or default values inside `metadata_entry`, as I'd have desired. There are some workarounds to process `metadata_entry {}` and distinguish that case from a malformed `metadata_entry` (ie: Missing sub-attributes). Also, `is_system` is a TypeBool sub-attribute, and the SDK doesn't consider tri-valued booleans. This means that an empty bool value is `false`, same as a real false value. The consequence of this is that `metadata_entry {}` will never be completely empty, so another fine-tune was done to the workaround for this.

In terms of current metadata, this PR unfortunately introduces the same behaviour for it:

```hcl
# The only way to delete metadata:
metadata = {}
```

But as it's a key/value map is much simpler and doesn't require any workaround. Just a mention in the docs.

### Tests

- [x] Acceptance tests pass
- [x] Binary tests pass
- [x] Upgrade tests pass (#927)

### CODE REVIEW

**Comment by adambarreiro** [comment]
I'm currently investigating a couple of failing tests after running the full acceptance suite:

- TestAccVcdCatalogAndItemDatasource
- TestAccVcdCatalogMediaBasic
- TestAccVcdCatalog
- TestAccVcdIndependentDiskBasic
- TestAccVcdNetworkIsolatedV2Nsxt
- TestAccVcdNetworkIsolatedV2NsxtMigration
- TestAccVcdNetworkRoutedV2Nsxt
- TestAccVcdNetworkIsolatedStatic1
- TestAccVcdNetworkRoutedStatic1
- TestAccVcdNetworkDirect

All are related to the **Import** operation, so I must have introduced a flaw there.

**Comment by adambarreiro** [comment]
Checking some test failures:

- TestAccVcdDistributedFirewall
- TestAccVcdDistributedFirewallVCD10_3_2
- TestAccVcdNsxtDynamicSecurityGroupVdcGroupCriteriaWithVms
- TestAccVcdNsxtRouteAdvertisementVdcGroup
- TestAccVcdNsxtSecurityGroupOwnerVdcGroup
- TestAccVcdNsxVdcGroupResources

All have some quirk with deprecated `metadata` field.

File: vcd/metadata_common_test.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+func stateDumper() resource.TestCheckFunc {
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/metadata_common_test.go]: I knew it would happen. I've been playing with it a lot and finally I committed it.

File: vcd/structure.go [hunk 2]
```diff
@@ -200,17 +200,19 @@ func contains(sliceToSearch []string, searched string) bool {
...
-// createOrUpdateOrgMetadata creates or updates metadata entries for the given resource and attribute name
+// createOrUpdateMetadata creates or updates metadata entries for the given resource and attribute name
+// TODO: This function implementation should be replaced with the implementation of `createOrUpdateMetadataInVcd`
+// once "metadata" field is removed.
 func createOrUpdateMetadata(d *schema.ResourceData, resource metadataCompatible, attributeName string) error {
-	if d.HasChange(attributeName) {
+	// We invoke the new "metadata_entry" metadata creation here to have it centralized and reduce duplication.
+	// Ideally, once "metadata" is removed in a new major, the implementation of `createOrUpdateMetadataInVcd` should
+	// just go here in the `createOrUpdateMetadata` body.
+	err := createOrUpdateMetadataInVcd(d, resource)
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/structure.go:210]: Nice, changed

File: vcd/structure.go [hunk 3]
```diff
@@ -200,17 +200,19 @@ func contains(sliceToSearch []string, searched string) bool {
...
-}
-
-// createOrUpdateOrgMetadata creates or updates metadata entries for the given resource and attribute name
+// createOrUpdateMetadata creates or updates metadata entries for the given resource and attribute name
+// TODO: This function implementation should be replaced with the implementation of `createOrUpdateMetadataInVcd`
+// once "metadata" field is removed.
 func createOrUpdateMetadata(d *schema.ResourceData, resource metadataCompatible, attributeName string) error {
-	if d.HasChange(attributeName) {
+	// We invoke the new "metadata_entry" metadata creation here to have it centralized and reduce duplication.
+	// Ideally, once "metadata" is removed in a new major, the implementation of `createOrUpdateMetadataInVcd` should
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/structure.go]: Fixed, thanks!

File: vcd/metadata_common_test.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+// - Step 4:  Delete 1 metadata entry, the bool one
+// - Step 5:  Update the string and date metadata values
+// - Step 6:  Delete all of them
+// - Step 7:  (Sysadmin only) Create 2 entries with is_system=true (readonly and private user_access)
+// - Step 8:  (Sysadmin only) Update the hidden one
+// - Step 9:  (Sysadmin only) Delete all of them
+// - Step 10:  Check a malformed metadata entry
+// - Step 11: (Org user only) Check that specifying an is_system metadata entry with a tenant user gives an error
+// - Step 12+: Some extra tests for deprecated `metadata` attribute
+func testMetadataEntryCRUD(t *testing.T, resourceTemplate, resourceAddress, datasourceTemplate, datasourceAddress string, extraParams StringMap) {
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply vcd/metadata_common_test.go:35]: Added `Taint` to the second step. That should force re-creation.

[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/metadata_common_test.go:35]: I have checked it and this works.
IDs of an entity get changed between steps 1 and 2 which means they were recreated.

File: website/docs/r/catalog_item.html.markdown [hunk 5]
```diff
@@ -48,7 +49,49 @@ The following arguments are supported:
 * `upload_piece_size` - (Optional) - Size in MB for splitting upload size. It can possibly impact upload performance. Default 1MB.
 * `show_upload_progress` - (Optional) - Default false. Allows seeing upload progress. (See note below)
 * `metadata` - (Optional; *v2.5+*) Key value map of metadata to assign to the associated vApp Template
-* `catalog_item_metadata` - (Optional; *v3.7+*) Key value map of metadata to assign to the Catalog Item
+* `metadata_entry` - (Optional; *v3.8+*) A set of metadata entries to assign to the Catalog Item. See [Metadata](#metadata) section for details.
```
[the remark this answers is not in the document]
Comment (adambarreiro) [reply website/docs/r/catalog_item.html.markdown:52]: To be precise:
- `metadata` will **remain** as it's vApp Template metadata
- `catalog_item_metadata` is marked as **deprecated** in favour of `metadata_entry`

**Review by Didainius — APPROVED** [review]

**Comment by adambarreiro** [comment]
Checking these test failures, related to Computed `metadata`:

TestAccVcdDistributedFirewall
TestAccVcdDistributedFirewallVCD10_3_2
TestAccVcdNsxtDynamicSecurityGroupVdcGroupCriteriaWithVms
TestAccVcdNsxtRouteAdvertisementVdcGroup
TestAccVcdNsxtSecurityGroupOwnerVdcGroup
TestAccVcdNsxVdcGroupResources

File: website/docs/r/network_routed_v2.html.markdown [hunk 6]
```diff
@@ -129,6 +130,55 @@ Static IP Pools support the following attributes:
 * `start_address` - (Required) The first address in the IP Range
 * `end_address` - (Required) The final address in the IP Range
 
+<a id="metadata"></a>
+## Metadata
```
Comment (lvirbalas) [inline website/docs/r/network_routed_v2.html.markdown:134]: Just thinking, maybe there's a way to do "includes" in these docs, so we don't have to repeat the same section many times in different files?

File: go.mod [hunk 7]
```diff
@@ -58,3 +58,5 @@ require (
 	google.golang.org/grpc v1.48.0 // indirect
```
Comment (lvirbalas) [inline go.mod:58]: This is unrelated to this file. The question is where is the change log? :)

File: website/docs/r/network_routed_v2.html.markdown (hunk 6 above)
Comment (adambarreiro) [reply website/docs/r/network_routed_v2.html.markdown:134]: I think there's no built-in tool in markdown to do that, but we could do a pre-processing tool, pretty much like we do with the changelog, something like a template. We can discuss with @Didainius @dataclouder. I also felt the pain when doing this huge copypaste but as it was some kind of exception I didn't explore this option (also it would be out of scope of this PR).

I also thought about just linking to a new "metadata.md" under `guides`, but as it's also quite small and not really a guide, I also discarded this option.

Let's hear your thoughts!

File: go.mod (hunk 7 above)
Comment (adambarreiro) [reply go.mod:58]: Unbelievably it's not there. I thought I added it but it was in the govcd PR. I'll add it.

Thanks for noticing!

File: website/docs/r/catalog_item.html.markdown (hunk 5 above)
[the remark this answers is not in the document]
Comment (lvirbalas) [reply website/docs/r/catalog_item.html.markdown:52]: > metadata will **remain** as it's vApp Template metadata

Oh, that's confusing :) Would it be possible to elaborate/add a sentence to the above doc line on `metadata` to clear this up? Or add a paragraph to the below "Metadata" section. What I'm looking for is a two item bulleted list explaining when to use `metadata` and when to use `metadata_entry` in this peculiar vApp Template/Catalog Item use case.

Thanks!

File: go.mod (hunk 7 above)
Comment (adambarreiro) [reply go.mod:58]: Fixed, thanks!

File: website/docs/r/network_routed_v2.html.markdown (hunk 6 above)
Comment (lvirbalas) [reply website/docs/r/network_routed_v2.html.markdown:134]: Thanks. It's fine for this PR, if there's no built-in way to do it.

**Review by lvirbalas — APPROVED** [review]

File: website/docs/r/catalog_item.html.markdown (hunk 5 above)
[the remark this answers is not in the document]
Comment (dataclouder) [reply website/docs/r/catalog_item.html.markdown:52]: It's legacy. Our original metadata in catalog item was actually using vApp Template metadata, which is convenient for users, as they can only set that kind of metadata in the UI. Then we added the metadata for catalog item itself, which could be different from the vApp Template's, and we called it `catalog_item_metadata`.

**Review by dataclouder — APPROVED** [review]

File: website/docs/r/catalog_item.html.markdown (hunk 5 above)
[the remark this answers is not in the document]
Comment (adambarreiro) [reply website/docs/r/catalog_item.html.markdown:52]: Sorry @lvirbalas, I missed your reply, but @dataclouder explained it perfectly!

[the remark this answers is not in the document]
Comment (adambarreiro) [reply website/docs/r/catalog_item.html.markdown:52]: I added a note (`->`) with an extra explanation, below the three attributes (you can check how it looks using the [preview tool](https://registry.terraform.io/tools/doc-preview)).
